src/merkle/store.rs

```diff
@@ -0,0 +1,375 @@
+//! An in-memory data store for Merkle-lized data
+//!
+//! This is a in memory data store for Merkle trees, this store allows all the nodes of a tree
+//! (leaves or internal) to live as long as necessary and without duplication, this allows the
+//! implementation of efficient persistent data structures
+use super::{
+    BTreeMap, Direction, MerkleError, MerklePath, MerkleTree, NodeIndex, Rpo256, RpoDigest,
+    SimpleSmt, Vec, Word,
+};
+
+// FIX: This is a temporary work around for the lack of Ord in BaseElement
+#[repr(transparent)]
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+struct WordWithOrd {
+    word: Word,
+}
+
+impl PartialOrd for WordWithOrd {
+    fn partial_cmp(&self, other: &Self) -> Option<core::cmp::Ordering> {
+        let self_ints = self.word.iter().map(|e| e.inner());
+        let other_ints = other.word.iter().map(|e| e.inner());
+
+        self_ints.partial_cmp(other_ints)
+    }
+}
+
+impl Ord for WordWithOrd {
+    fn cmp(&self, other: &Self) -> core::cmp::Ordering {
+        let self_ints = self.word.iter().map(|e| e.inner());
+        let other_ints = other.word.iter().map(|e| e.inner());
+
+        self_ints.cmp(other_ints)
+    }
+}
+
+impl From<Word> for WordWithOrd {
+    fn from(word: Word) -> Self {
+        WordWithOrd { word }
+    }
+}
+
+impl From<WordWithOrd> for Word {
+    fn from(word: WordWithOrd) -> Self {
+        word.word
+    }
+}
+
+impl From<RpoDigest> for WordWithOrd {
+    fn from(digest: RpoDigest) -> Self {
+        WordWithOrd {
+            word: digest.into(),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct Node {
+    left: WordWithOrd,
+    right: WordWithOrd,
+}
+
+pub struct MerkleStore {
+    nodes: BTreeMap<WordWithOrd, Node>,
+}
```

Comment:
What is the reason for using `WordWithOrd` here instead of `RpoDigest` (or its alias `Digest`)? I think `Digest` implements `Ord` already and we can convert between `Word` and `Digest` pretty easily and efficiently.

Reply:
That's true, we do have `Ord` there. We are using Montgomery form tho. I'm wondering if that is secure or we might step into an edge case someday 🤔

@Al-Kindi-0

Reply:
> What is the reason for using WordWithOrd here instead of RpoDigest (or its alias Digest)?

I didn't know `RpoDigest` has `Ord` implemented 😅.

I switched to Word because it was easier to have a single type instead of mixing multiple, let me know what you guys prefer

Reply:
I'd prefer to use `Digest` as this is what we actually store. `Word` is just something we use for the external interface as it would make it easier to interact with the VM.

Reply:
changed. the resulting code is much shorter :)